flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -149,6 +149,47 @@ void assignExclusiveSegments() throws IOException {
 		}
 	}
 
+	@Override
+	public void initializeState(ChannelStateReader reader) throws IOException, InterruptedException {
+		numRequiredBuffers = initialCredit + inputGate.getBufferPool().getMaxNumberOfMemorySegments();
+		unannouncedCredit.set(initialCredit);
+
+		while (true) {
+			Buffer buffer;
+			synchronized (bufferQueue) {
+				buffer = bufferQueue.takeBuffer();
+				if (buffer == null) {
+					if (isReleased()) {
+						return;
+					}
+
+					if (!isWaitingForFloatingBuffers) {
+						buffer = inputGate.getBufferPool().requestBuffer();
+						if (buffer == null) {
+							inputGate.getBufferProvider().addBufferListener(this);
+							isWaitingForFloatingBuffers = true;
+						}
+					}
+				}
+			}
+
+			if (buffer == null) {
+				wait(10);
```

Comment:
Is someone notifing this thread to wake up? Can not we wait on the buffer in another way?

Reply:
TBH the current wait way is also not in my favor as I mentioned as unsatisfied points, because how long time to wait might be an issue.

I also considered other ways, but the key problem is that we can not exit this runnable to let the thread execute other runables in advance, so it seems somehow as blocking way here. In contrast, the previous floating request from netty thread was non-blocking way, so we do not need the mechanism of `notifyAll`  inside `RemoteInputChannel#recycle` and `RemoteInputChannel#notifyBufferAvailable`.

Maybe it is possible to add `notifyAll` in above two methods for compatible with the new process. WDYT?

Reply:
Ok I see. We can not block the shared executor for unspecified long time. This method would have to be rewritten in a such way, that once buffers are not available, it exits, and re-enqueuing action is done once more buffers are ready.

If we spawned our custom thread for unspilling, we could block it however we want, but in that case we would also need a mechanism to wake it up once more buffers are ready, instead of sleeping for fixed amount of time.

Reply:
Yes, we have the same issue in another option of `SpilledInputChannel`/`SpilledInputGate`.

In conclusion, there may be several solutions:

- `wait()` if unavailable buffers: block the unspilling thread always and it fits our current requirements. We should not exit current channel to switch another channel temporarily, which might bring random IO. And it requires the wakeup mechanism when buffer available again.

- `wait(timeout)` if unavailable buffers: more or less the same with above `wait()`, but wakeup mechanism is not a mandatory, can be regarded as  somehow improvement to wakeup eerily. 

- unblocking way: terminate the current channel and unspill another channel with available buffers. It would bring random IO as mentioned above and not the current suggestion.

Based on the current situation with custom thread for unspilling, and allow only one thread to unspill channel one by one to avoid random IO, so I choose the unblocking option 1 or 2. 

Regarding the option 2, wakeup is not necessary, then we do not need to touch the previous processes `RemoteInputChannel#recycle` and `RemoteInputChannel#notifyBufferAvailable`. If we want to add the wakeup mechanism, option 1 also makes sense.

Reply:
Oh, I missed that we are already using a separate thread pool. In that case we don't need the not blocking option now.

It would be still good to get rid of timed wait in favour of some notifications though.